internal/events/batch_pin_complete.go

```diff
@@ -48,24 +46,33 @@ func (em *eventManager) BatchPinComplete(bi blockchain.Plugin, batchPin *blockch
 	}()
 	log.L(em.ctx).Tracef("BatchPinComplete batch=%s info: %+v", batchPin.BatchID, batchPin.Event.Info)
 
-	if batchPin.BatchPayloadRef != "" {
-		return em.handleBroadcastPinComplete(batchPin, signingKey)
-	}
-	return em.handlePrivatePinComplete(batchPin, signingKey)
-}
-
-func (em *eventManager) handlePrivatePinComplete(batchPin *blockchain.BatchPin, signingKey *fftypes.VerifierRef) error {
 	// Here we simple record all the pins as parked, and emit an event for the aggregator
 	// to check whether the messages in the batch have been written.
-	return em.retry.Do(em.ctx, "persist private batch pins", func(attempt int) (bool, error) {
+	return em.retry.Do(em.ctx, "persist batch pins", func(attempt int) (bool, error) {
 		// We process the batch into the DB as a single transaction (if transactions are supported), both for
 		// efficiency and to minimize the chance of duplicates (although at-least-once delivery is the core model)
 		err := em.database.RunAsGroup(em.ctx, func(ctx context.Context) error {
-			err := em.persistBatchTransaction(ctx, batchPin)
-			if err == nil {
-				err = em.persistContexts(ctx, batchPin, signingKey, true)
+			chainEvent := buildBlockchainEvent(batchPin.Namespace, nil, &batchPin.Event, &fftypes.TransactionRef{
+				Type: fftypes.TransactionTypeBatchPin,
+				ID:   batchPin.TransactionID,
+			})
+			if err := em.persistBlockchainEvent(ctx, chainEvent); err != nil {
+				return err
 			}
-			return err
+			if err := em.persistBatchTransaction(ctx, batchPin); err != nil {
```

Comment:
Minor detail, but maybe we should persist the transaction before the event, since the event refers to the transaction...